### ISSUE TITLE
Use of RandomizedSearchCV with Nolearn?

### ISSUE THREAD

**Issue by BrianMiner** [dnouri/nolearn#95]
Is there a known issue / limitation in using nolearn (lasagne) with scikit's randomizedsearchCV to tune parameters? When I  run the following snipet, I get results that appear that there is a problem. It appears that the cv =3 causes three identical runs of the code simultaneously and the result are nan values for the loss or impossibly large values (50+ digits). 
```
 param_dist = {"update_learning_rate": sp_rand(0.001,0.01),
                  "dense0_num_units": sp_randint(100,300),
                  "dense1_num_units": sp_randint(100,300),
                  "dropout0_p":  sp_rand(0.1,0.75),
                  "dropout1_p":  sp_rand(0.1,0.75),
                  "update_momentum" : sp_rand(0.9,0.99),
                  "max_epochs": [50,100,150,200]
                   
                  }

    
    layers0 = [('input', InputLayer),
           ('dense0', DenseLayer),
           ('dropout', DropoutLayer),  
           ('dense1', DenseLayer),
           ('dropout1', DropoutLayer),  
           ('output', DenseLayer)]
    
    net0 = NeuralNet(layers=layers0,
                input_shape=(None, train_X.shape[1]), 
                dense0_nonlinearity=rectify,
                 dense1_nonlinearity=rectify,
                 output_num_units=4,
                 output_nonlinearity=softmax,
                 objective_loss_function =categorical_crossentropy,
                 update=nesterov_momentum,
                 eval_size=0.0,
                 verbose=1
                   )
     
random_search = RandomizedSearchCV(net0, param_distributions=param_dist, n_iter=50, scoring =kappa , cv=3, n_jobs=1)
random_search.fit(train_X, train_y.values)
```

*Output*
```
input                 (None, 313)             produces     313 outputs
  dense0                (None, 100)             produces     100 outputs
  dropout               (None, 100)             produces     100 outputs
  dense1                (None, 253)             produces     253 outputs
  dropout1              (None, 253)             produces     253 outputs
  output                (None, 4)               produces       4 outputs
# Neural Network with 57969 learnable parameters

## Layer information

|   # | name     |   size |
|----:|:---------|-------:|
|   0 | input    |    313 |
|   1 | dense0   |    100 |
|   2 | dropout  |    100 |
|   3 | dense1   |    253 |
|   4 | dropout1 |    253 |
|   5 | output   |      4 |

# Neural Network with 57969 learnable parameters

## Layer information

|   # | name     |   size |
|----:|:---------|-------:|
|   0 | input    |    313 |
|   1 | dense0   |    100 |
|   2 | dropout  |    100 |
|   3 | dense1   |    253 |
|   4 | dropout1 |    253 |
|   5 | output   |      4 |

# Neural Network with 57969 learnable parameters

## Layer information

|   # | name     |   size |
|----:|:---------|-------:|
|   0 | input    |    313 |
|   1 | dense0   |    100 |
|   2 | dropout  |    100 |
|   3 | dense1   |    253 |
|   4 | dropout1 |    253 |
|   5| output   |      4 |

  epoch    train loss    valid loss    train/val    valid acc  dur
-------  ------------  ------------  -----------  -----------  -----
      1           nan           nan          nan          nan  0.19s
  epoch    train loss    valid loss    train/val    valid acc  dur
-------  ------------  ------------  -----------  -----------  -----
      1           nan           nan          nan          nan  0.19s
  epoch    train loss    valid loss    train/val    valid acc  dur
-------  ------------  ------------  -----------  -----------  -----
      1           nan           nan          nan          nan  0.19s
      2           nan           nan          nan          nan  0.18s
      2           nan           nan          nan          nan  0.18s
      2           nan           nan          nan          nan  0.18s
      3           nan           nan          nan          nan  0.18s
      3           nan           nan          nan          nan  0.18s
      3           nan           nan          nan          nan  0.18s
      4           nan           nan          nan          nan  0.18s
      4           nan           nan          nan          nan  0.18s
      4           nan           nan          nan          nan  0.18s
      5           nan           nan          nan          nan  0.18s
      5           nan           nan          nan          nan  0.18s
      5           nan           nan          nan          nan  0.18s
      6           nan           nan          nan          nan  0.18s
      6           nan           nan          nan          nan  0.18s
      6           nan           nan          nan          nan  0.18s
      7           nan           nan          nan          nan  0.19s
      7           nan           nan          nan          nan  0.19s
      7           nan           nan          nan          nan  0.19s
      8           nan           nan          nan          nan  0.18s
      8           nan           nan          nan          nan  0.18s
      8           nan           nan          nan          nan  0.18s
      9           nan           nan          nan          nan  0.18s
      9           nan           nan          nan          nan  0.18s
      9           nan           nan          nan          nan  0.18s
     10           nan           nan          nan          nan  0.19s
     10           nan           nan          nan          nan  0.19s
     10           nan           nan          nan          nan  0.19s
     11           nan           nan          nan          nan  0.18s
     11           nan           nan          nan          nan  0.18s
     11           nan           nan          nan          nan  0.18s
     12           nan           nan          nan          nan  0.19s
     12           nan           nan          nan          nan  0.19s
     12           nan           nan          nan          nan  0.19s
     13           nan           nan          nan          nan  0.18s
     13           nan           nan          nan          nan  0.18s
     13           nan           nan          nan          nan  0.18s
     14           nan           nan          nan          nan  0.19s
     14           nan           nan          nan          nan  0.19s
     14           nan           nan          nan          nan  0.19s
     15           nan           nan          nan          nan  0.19s
     15           nan           nan          nan          nan  0.19s
     15           nan           nan          nan          nan  0.19s
     16           nan           nan          nan          nan  0.18s
     16           nan           nan          nan          nan  0.18s
     16           nan           nan          nan          nan  0.18s
     17           nan           nan          nan          nan  0.18s
     17           nan           nan          nan          nan  0.18s
     17           nan           nan          nan          nan  0.18s 


Or if there is a value it looks like this;

```
  epoch    train loss    valid loss    train/val    valid acc  dur
-------  ------------  ------------  -----------  -----------  -----
      1       1.21497           nan          nan          nan  0.21s
  epoch    train loss    valid loss    train/val    valid acc  dur
-------  ------------  ------------  -----------  -----------  -----
      1       1.21497           nan          nan          nan  0.21s
  epoch    train loss    valid loss    train/val    valid acc  dur
-------  ------------  ------------  -----------  -----------  -----
      1       1.21497           nan          nan          nan  0.21s
      2       1.07145           nan          nan          nan  0.21s
      2       1.07145           nan          nan          nan  0.21s
      2       1.07145           nan          nan          nan  0.21s
      3       1.00561           nan          nan          nan  0.21s
      3       1.00561           nan          nan          nan  0.21s
      3       1.00561           nan          nan          nan  0.21s
      4       0.95005           nan          nan          nan  0.21s
      4       0.95005           nan          nan          nan  0.21s
      4       0.95005           nan          nan          nan  0.21s
      5       0.90978           nan          nan          nan  0.21s
      5       0.90978           nan          nan          nan  0.21s
      5       0.90978           nan          nan          nan  0.21s
      6       0.87553           nan          nan          nan  0.21s
      6       0.87553           nan          nan          nan  0.21s
      6       0.87553           nan          nan          nan  0.21s
      7       0.84362           nan          nan          nan  0.21s
      7       0.84362           nan          nan          nan  0.21s
      7       0.84362           nan          nan          nan  0.21s
      8       0.83171           nan          nan          nan  0.21s
      8       0.83171           nan          nan          nan  0.21s
      8       0.83171           nan          nan          nan  0.21s
      9       0.80696           nan          nan          nan  0.21s
      9       0.80696           nan          nan          nan  0.21s
      9       0.80696           nan          nan          nan  0.21s

```






```

**Comment by BrianMiner** [comment]
ping?

**Comment by asantinc** [comment]
I'm not sure if the 3 runs are occurring simultanously. What I have observed is that the NeuralNetwork [gets cloned] (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/grid_search.py#L538-538) by GridSearch. Every time that happens a new instance of PrintLog [gets appended] (https://github.com/dnouri/nolearn/blob/master/nolearn/lasagne/base.py#L145-150) to the list of functions to be executed at the end of each epoch

**Comment by BrianMiner** [comment]
I am going to close this and open a new issue with a few more examples
